cephbootstrap: run bootstrap with --apply-spec

smithfarm opened the pull request:
As reported in the downstream bug cited below, "cephadm bootstrap"
is now deploying 5 MONs and 2 MGRs by default.

The expectation is that "ceph-salt apply" will cause one (and only one)
MON and one (and only one) MGR to be deployed, and that both the MON and
the MGR will be deployed on the node specified by the user as the
bootstrap node.

Fixes: https://bugzilla.suse.com/show_bug.cgi?id=1174422
Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm committed:
cephbootstrap: put bootstrap options in alphabetical order

Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm committed:
cephbootstrap: run bootstrap with --apply-spec

As reported in the downstream bug cited below, "cephadm bootstrap"
is now deploying 5 MONs and 2 MGRs by default.

The expectation is that "ceph-salt apply" will cause one (and only one)
MON and one (and only one) MGR to be deployed, and that both the MON and
the MGR will be deployed on the node specified by the user as the
bootstrap node.

Fixes: https://bugzilla.suse.com/show_bug.cgi?id=1174422
Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm commented:
c3bf1ebf4c07b6f3ec6a2da671a3adc590c8f9f4 started installing "ceph-base" RPM on minions before the RPM was added to the product Media: added temporary revert for testing

ricardoasmarques commented on ceph-salt-formula/salt/ceph-salt/apply/cephbootstrap.sls:
```diff
@@ -6,6 +6,22 @@
+        echo -en "service_name: mgr\n" >> /tmp/static_bootstrap.yaml
+        echo -en "placement:\n" >> /tmp/static_bootstrap.yaml
+        echo -en "    hosts:\n" >> /tmp/static_bootstrap.yaml
+        echo -en "        - {{ pillar['ceph-salt']['bootstrap_node_hostname'] }}\n" >> /tmp/static_bootstrap.yaml
```
Instead of adding a new pillar item, can we simply read `host` grain like we are doing in https://github.com/ceph/ceph-salt/blob/master/ceph-salt-formula/salt/ceph-salt/apply/cephorch.sls#L8 ?

Note that this is running in the bootstrap minion, so it should work.

ricardoasmarques commented on ceph-salt-formula/salt/ceph-salt/apply/cephbootstrap.sls:
```diff
@@ -6,6 +6,22 @@
+        echo -en "service_name: mon\n" >> /tmp/static_bootstrap.yaml
+        echo -en "placement:\n" >> /tmp/static_bootstrap.yaml
+        echo -en "    hosts:\n" >> /tmp/static_bootstrap.yaml
+        echo -en "        - {{ pillar['ceph-salt']['bootstrap_node_hostname'] }}\n" >> /tmp/static_bootstrap.yaml
```
Perhaps `{{ grains['host'] }}`?

ricardoasmarques commented on ceph_salt/config_shell.py:
```diff
@@ -96,6 +96,7 @@ def save(self, value):
             raise MinionDoesNotExistInConfiguration(value)
         node = CephNodeManager.ceph_salt_nodes()[value]
         PillarManager.set('ceph-salt:bootstrap_mon_ip', node.public_ip)
+        PillarManager.set('ceph-salt:bootstrap_node_hostname', node.hostname)
```
Hopefully, we can remove this.

smithfarm committed:
cephbootstrap: use Jinja var for bootstrap ceph conf tmpfile

This hopefully makes the code a little bit more maintainable.

Signed-off-by: Nathan Cutler <ncutler@suse.com>

sebastian-philipp commented:
lgtm

smithfarm committed:
cephbootstrap: put Mon ID in placement YAML

Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm commented:
So, I added a commit to specify the Mon IP in the placement YAML. It didn't make sense to me to specify the same piece of information in two places, so I dropped the `--mon-ip` option. However, when invoked without `--mon-ip` or `--mon-addrv`, `cephadm bootstrap` barfs out a Traceback:

```
Traceback (most recent call last):
  File "/usr/sbin/cephadm", line 5046, in <module>
    r = args.func()
  File "/usr/sbin/cephadm", line 1214, in _default_image
    return func()
  File "/usr/sbin/cephadm", line 2537, in command_bootstrap
    raise Error('must specify --mon-ip or --mon-addrv')
__main__.Error: must specify --mon-ip or --mon-addrv
```

Opened https://tracker.ceph.com/issues/46782 to track this upstream.

smithfarm commented:
@ricardoasmarques I just tested this again and it works for me.

ricardoasmarques commented:
> So, I added a commit to specify the Mon IP in the placement YAML. It didn't make sense to me to specify the same piece of information in two places, so I dropped the `--mon-ip` option. However, when invoked without `--mon-ip` or `--mon-addrv`, `cephadm bootstrap` barfs out a Traceback:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/sbin/cephadm", line 5046, in <module>
>     r = args.func()
>   File "/usr/sbin/cephadm", line 1214, in _default_image
>     return func()
>   File "/usr/sbin/cephadm", line 2537, in command_bootstrap
>     raise Error('must specify --mon-ip or --mon-addrv')
> __main__.Error: must specify --mon-ip or --mon-addrv
> ```
> 
> Opened https://tracker.ceph.com/issues/46782 to track this upstream.

@sebastian-philipp is your approval still valid now that we are specifying the Mon IP in the placement YAML?

ricardoasmarques reviewed (approved):
lgtm

sebastian-philipp commented:
we have to specify the `--mon-ip` here. bootstrap works like so:

1. create the first mon using `--mon-ip`
1. create the first mgr
1. configure dashboard etc...
1. `ceph orch apply -i spec.yaml`
1. `mgr/cephadm` will follow the instructions that were passed in the spec.

Which means, we're looking at the spec **much** later. Meaning after bootstrapping the first mon. There is no way for now to look into the spec.yaml at this point and extract the mon ip.

sebastian-philipp commented:
lgtm

smithfarm commented:
@sebastian-philipp Ah! Now I understand what is missing. If we had a `--no-apply-spec` option, we could restore the original behavior of `cephadm bootstrap`, which was to create 1 MON and 1 MGR on the node where it is running.

Then we wouldn't need to bend over backwards sending `cephadm bootstrap` a spec file which merely states what has already done, in order to prevent `cephadm bootstrap` from "helpfully" creating 5 MONs and 2 MGRs.

What do you think?

smithfarm commented:
Opened https://tracker.ceph.com/issues/46801 to track the issue, so we don't block this PR.

sebastian-philipp commented:
> @sebastian-philipp Ah! Now I understand what is missing. If we had a `--no-apply-spec` option, we could restore the original behavior of `cephadm bootstrap`, which was to create 1 MON and 1 MGR on the node where it is running.

The `--no-apply-spec` is called `--orphan-initial-daemons`, but you really don't want to do that, except if you're Teuthology, as it leaves the deplyed daemons orphaned and scheduled for potential removal


> 
> Then we wouldn't need to bend over backwards sending `cephadm bootstrap` a spec file which merely states what has already done, in order to prevent `cephadm bootstrap` from "helpfully" creating 5 MONs and 2 MGRs.
> 
> What do you think?

smithfarm commented:
> The `--no-apply-spec` is called `--orphan-initial-daemons`, but you really don't want to do that, except if you're Teuthology, as it leaves the deplyed daemons orphaned and scheduled for potential removal

In that case, it sounds like we need a "--no-apply-spec" that is just like `--orphan-initial-daemons`, except it doesn't orphan the initial daemons :-)
